### PR TITLE
Add support for app.enhance

### LINKED ISSUE #90: Investigate API for composition within DI system
I had some ideas surrounding composition within the DI system that I wanted to get down. This is purely a brain dump and not a final interface.

A classic example of where this could be useful is the `fetch` token. We already have an example of a plugin which requires a `fetch` implementation, and provides a new `fetch` implementation - `fusion-plugin-csrf-protection`. We solved this problem through token aliasing, which works well enough but I think it can be a bit difficult to understand at first.

Lets say we had three plugins that wanted to register on the `FetchToken` (polyfill, routePrefix, csrf). Using token aliasing, we would have something like the following:

```js
const BaseFetchToken = createToken('BaseFetch');
const PrefixFetchToken = createToken('PrefixFetch');
app.register(BaseFetchToken, window.fetch);
app.register(PrefixFetchToken, RoutePrefixFetchPlugin).alias(FetchToken, BaseFetchToken);
app.register(FetchToken, CsrfPlugin).alias(FetchToken, PrefixFetchToken);
```

This is fine, but I think it would be better if we didn't have to keep track of all these different tokens. Using composition it would look like this:

```js
app.register(FetchToken, window.fetch);
app.registerCompose(FetchToken, RoutePrefixComposition);
app.registerCompose(FetchToken, CsrfComposition);
```

Now with less hand waiving...

```js
app.register(FetchToken, window.fetch);
// add routePrefix
app.registerCompose(FetchToken, (fetch) => {
  return (url, options) => {
    return fetch(window.routePrefix + url, options);
  }
});
// add csrf
app.registerCompose(FetchToken, (fetch) => {
  return createPlugin({...});
});
```

The first argument to `registerCompose` is the Token to register and compose with, and the second argument is function that takes the current resolved `provides` of the composed token, (in this case the implementation of `fetch`) and returns either a `plugin` or a new composed value directly. In the case of returning a `plugin`, the value that will be composed will be the result of the plugin's `provides` function.

Thoughts?

### PR DESCRIPTION
**Pull request by ganemone**
Fixes #90

### CODE REVIEW

File: src/__tests__/dependency-resolution.js [hunk 1]
```diff
@@ -331,7 +331,8 @@ tape('dependency registration with optional dependency', t => {
     },
   });
   app.register(TokenC, PluginWithOptionalDependency);
-
+  // eslint-disable-next-line
+  debugger;
```
Comment (dennisgl) [inline src/__tests__/dependency-resolution.js]: redundant statement

Comment (KevinGrandon) [inline src/__tests__/dependency-resolution.js]: Seems like we should remove this?

Comment (ganemone) [reply src/__tests__/dependency-resolution.js]: good catch!

**Comment by ganemone** [comment]
!merge

**Pull request merged**